Add a curves module and a StandardFeatures transformer to extract features from them

gtauzin opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the guidelines for contributing: https://giotto-ai.github.io/gtda-docs/latest/contributing/#guidelines
-->

**Reference issues/PRs**
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests.

Pull requests which are not related to open issues may be rejected!
If suggesting a new feature or change, please discuss it in an issue first.
If fixing a bug, there should be an issue describing it with steps to reproduce.
-->

**Types of changes**
<!--
What types of changes does your code introduce? Put an `x` in all the boxes that apply.
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

**Description**
<!--
Describe your changes in detail.
-->
- Adds a `curves` module to process and extract features from Betti, Landscape and Silhouette
- Adds a `StandardFeature` transformer that can return a variety of functions on those curves as features
- Make `PersistentLandscape` return a 3D array, i.e. a multicurve that can be processed by the `curves` transformers too.

**Screenshots (if appropriate)**

**Any other comments?**

**Checklist**
<!--
Go over all the following points, and put an `x` in all the boxes that apply. 
If you're unsure about any of these, don't hesitate to ask. We're here to help!
-->
- [x] I have read the [guidelines for contributing](https://giotto-ai.github.io/gtda-docs/latest/contributing/#guidelines).
- [x] My code follows the code style of this project. I used `flake8` to check my Python changes.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [ ] All new and existing tests passed. I used `pytest` to check this on Python tests.

<!--
We value all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.

Thanks for contributing!
-->

Commit:
Add curves submodule with StandardFeature

Signed-off-by: Guillaume Tauzin <guillaume.tauzin@inait.ai>

Commit:
Fix linting

Signed-off-by: Guillaume Tauzin <guillaume.tauzin@inait.ai>

wreise commented on gtda/curves/features.py:
```diff
@@ -0,0 +1,116 @@
+class StandardFeature(BaseEstimator, TransformerMixin):
+    """Computes standard features of multi-channels curves.
+
+    Given a multi curve applies any function to extract features from it.
```
What about "(...) multi-channel curve (...)" ?

wreise commented on gtda/curves/features.py:
```diff
@@ -0,0 +1,116 @@
+            self.effective_function_params_ = self.function_params.copy()
+        validate_params(
+            self.effective_function_params_,
+            _AVAILABLE_FUNCTIONS[self.function])
```
That might not work when `self.function` is a callable.

wreise commented on gtda/curves/features.py:
```diff
@@ -0,0 +1,116 @@
+
+        Returns
+        -------
+        Xt : ndarray of shape (n_samples, n_features)
```
Is `n_features` the dimension `self.function(x)`, or maybe it is the same as `n_channels`?

wreise commented on gtda/curves/features.py:
```diff
@@ -0,0 +1,116 @@
+    function : str or callable, optional, default: ``max``
+        Function to transform a multi channel curve into features. Implemented
+        functions are [``'argmin'``, ``'argmax'``, ``'min'``, ``'max'``,
+        ``'mean'``, ``'std'``, ``'median'``, ``'average'``]
```
So, in general, suppose I have a multichannel curve `S`, is f applied on each channel of S, or on S as a whole?

gtauzin replied:
I tried to make that a bit more clear. Let me know what you think.

wreise reviewed (commented):
A few docstring questions and suggestions.

gtauzin committed:
Merge branch 'master' of github.com:giotto-ai/giotto-tda into curves

gtauzin committed:
Fix curves from @wreise comment, add 'flatten'

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Readd tests

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Fix docstrings

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Reshape output of landscape so that it's a multi channel curve

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Fix linting

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Fix plotting method and test

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Fix linting

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

gtauzin committed:
Ensure params are vlidated only if function is str

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

ulupo commented on gtda/curves/__init__.py:
```diff
@@ -0,0 +1,8 @@
+"""The module :mod:`gtda.features` implements transformers to postprocess
```
Should this be ``` :mod:`gtda.curves` ```?

wreise replied:
:+1:

gtauzin replied:
Too many PRs opened at the same time.

ulupo commented on gtda/curves/_functions.py:
```diff
@@ -0,0 +1,39 @@
+}
+
+_implemented_function_recipes = {
+    'flatten': lambda X, axis: np.reshape(X, (X.shape[0], -1)),
```
The `axis` argument seems to be unused.

gtauzin replied:
On purpose. All the other takes one, so instead of redefining all other function with axis=-1, I hardcoded it in this lambda.

ulupo replied:
Thanks for the explanation!

ulupo reviewed (changes requested):
I'm not sure I'm too happy with the breaking change with `PersistenceLandscape`, let me see whether I can find an alternative. I understand where the change is coming from, though ;)

wreise reviewed (commented):
Thanks, the docs are better :)

gtauzin committed:
Fix typo

Signed-off-by: Guillaume Tauzin <guillaumetauzin.ut@gmail.com>

ulupo committed:
Merge branch 'master' into curves

ulupo committed:
Merge branch 'master' into curves

ulupo committed:
Linting/license

ulupo committed:
Add license to curves/features.py

ulupo committed:
Shorter title for StandardFeature

ulupo committed:
Add curves to main init and to rst docs

ulupo committed:
Add Feature extraction subtitle in curves.rst

ulupo committed:
Split simplicial homology into undirected and directed subsections

ulupo committed:
Fix typo in validate_params

ulupo committed:
Add _validate_params method to allow for function input

ulupo committed:
Rename to StandardFeatures and refactor

- Allow lists/tuples of functions with None
- Remove 'flatten' (out of scope because of output shape)
- Improve tests
- Impose 3D input

Docs still WIP

ulupo committed:
Reinstate flatten but this time it is called identity

ulupo committed:
Test vector-valued functions

ulupo committed:
Cover case of mixtures of functions with different output shapes

ulupo committed:
Fix typo

ulupo committed:
Fix linting

ulupo committed:
Add error if transform channels are not the same as fit channels

ulupo committed:
Add test for behaviour when transform channels and fit channels differ

ulupo committed:
Remove inspect signature logic, extend function_params to lists of dict, improve tests

ulupo committed:
Fix errors and tests, remove dim=3 check, remove function_

ulupo committed:
Raise test coverage to 100%

ulupo committed:
Add random generator seed

ulupo committed:
Reinstate check of 3 dimensions

ulupo committed:
Write X explicitly

ulupo committed:
Slashes in comment

ulupo committed:
Silence DeprecationWarning from numpy array creation from ragged sequences

ulupo committed:
Fix use of warnings context manager

ulupo committed:
Use np.ones to avoid scikit-learn errors from large floats

ulupo committed:
Fix DeprecationWarning -> VisibleDeprecationWarning

gtauzin commented on gtda/curves/features.py:
```diff
@@ -0,0 +1,217 @@
+
+    Parameters
+    ----------
+    function : string or callable, optional, default: ``max``
```
I think it would be nice to have the possibility to have None here. None would return an empy feature array (n_samples, 0). It would be useful if you do not wish to get any features for a specific homology dimension.

ulupo replied:
Sorry, I'm a little confused! If homology dimension is a channel, isn't this the capability allowed by having `function` as a list with some Nones? Or is this a case of you looking before the docs were ready? XD

gtauzin replied:
Yes :)

gtauzin commented on gtda/curves/features.py:
```diff
@@ -0,0 +1,217 @@
+                     "in": tuple(_AVAILABLE_FUNCTIONS.keys()),
+                     "of": {"type": (str, FunctionType, type(None)),
+                            "in": tuple(_AVAILABLE_FUNCTIONS.keys())}},
+        "function_params": {"type": (dict, type(None), list, tuple)},
```
Is it useful to have it None here? As we don't modify it, the `effective_function_params_` attribute will be the same than `function_params` or a list of multiple copies of it, right?

ulupo replied:
This is as in your original design. I think it's ok to have it as it is maybe future-proof?

ulupo replied:
Ah no wait, there's a better reason. It must be None otherwise the default would break either the single function case or the list of functions case! E.g. if we commit to the default being {} then a user passing a list of functions will get an error.

gtauzin replied:
Ah, my bad then! :)
But that's because we treat it that way, right. We could actually allow empty dictionnaries to act as if they were None.
The problem is that right now, the `effective_function_params_` attribute that arises from this treatment, does never contains information it seems. Is that right?

I can also take care of it if you prefer.

ulupo replied:
> We could actually allow empty dictionnaries to act as if they were None.

In principle yes, but I feel it is slightly clearer to have the `None` to avoid confusions.

> The problem is that right now, the effective_function_params_ attribute that arises from this treatment, does never contains information it seems. Is that right?

Wasn't this also the case before? What has changed in your opinion?

gtauzin replied:
As I said it' my bad. It was there before. :)

I was also thinking, that `effective_function_` should be an attribute too! As the list of the actual function applied to each channel. In that sense, maybe `effective_function_params_` would make sense to have as an attribute as a list of dictionarry.

ulupo replied:
Yeah so I was very torn about `effective_function_` as I was wondering whether it should always be a tuple of functions (or `None`) even when `function` was a single function.  I think it is less confusing if `effective_function_` and `effective_function_params_` have always the same type (tuple).  Do you agree that I can:
- add `effective_function_` which is always a tuple of functions or `None` even if the user passes one of the strings or a single function;
- ensure `effective_function_params_` is also always a tuple of (possibly empty) dictionaries.

gtauzin replied:
I agree! Why not list btw, given than we also agreed on reducing the complexity of the parallel function with list indexing?

ulupo replied:
I try to be consistent throughout the library with list-like attributes, such as `homology_dimensions_`, to make them all tuples. I think mutable objects should be avoided when they are not meant to be manipulated but just inspected. And I think this is the case of these class attributes, no?

gtauzin replied:
Ok, I don't think that's the case all over the library, but why not!

ulupo replied:
Oh ok! I once went through all submodules to check for this, so if you find some inconsistencies please let me know ;)

gtauzin replied:
I trust you xD

gtauzin commented on gtda/curves/features.py:
```diff
@@ -0,0 +1,217 @@
+        """
+        check_is_fitted(self)
+        Xt = check_array(X, ensure_2d=False, allow_nd=True)
+        if Xt.ndim != 3:
```
For far away in the future: we should be able to pass detailed shape requirements to a `check_array` of our own.

gtauzin commented on gtda/curves/features.py:
```diff
@@ -0,0 +1,217 @@
+
+    Parameters
+    ----------
+    function : string or callable, optional, default: ``max``
```
I think it would be nice to have the possibility to have None here. None would mean that it returns an empty array (n_samples, 0) but if used in a list it could mean to skip feature generation for that specific channel.

gtauzin replied:
As far as I can see it is done in tuple/list which is more than enough, thanks!

gtauzin commented on gtda/curves/features.py:
```diff
@@ -0,0 +1,217 @@
+                     "in": tuple(_AVAILABLE_FUNCTIONS.keys()),
+                     "of": {"type": (str, FunctionType, type(None)),
+                            "in": tuple(_AVAILABLE_FUNCTIONS.keys())}},
+        "function_params": {"type": (dict, type(None), list, tuple)},
```
It seems to me that there is no need to have `function_params` as None as we would never need to modify it in `fit`. Here the corresponding attribute is always either `function_params` or list of `function_params` copies, right?

gtauzin commented on gtda/curves/features.py:
```diff
@@ -0,0 +1,217 @@
+
+        """
+        check_array(X, ensure_2d=False, allow_nd=True)
+        if X.ndim != 3:
```
For far away in the future, it would be nice to be able to specifiy details about the shape to a `check_array` of our own.

ulupo replied:
Agree! It is useful both here and in images for instance.

gtauzin commented on gtda/curves/features.py:
```diff
@@ -0,0 +1,217 @@
+                        tuple([{}] * self.n_channels_)
+                else:
+                    self.effective_function_params_ = \
+                        tuple(self.function_params.copy()
```
Let me try this xD
```
tuple([self.function_params.copy()] * self.n_channels_)
```

ulupo replied:
Aha! I thought this would be a bad trap because each entry in `[self.function_params.copy()] * self.n_channels_` refers to the same python object.
Try this:
```
d = {"spam": "eggs"}
t = tuple([d] * 2)
t[0].pop("spam")
```

gtauzin replied:
Yes, but we are not modifying them further. As long as the initial `function_params` can't be modified, I it's fine, no?

ulupo replied:
You are right about this, I was being over-cautious here!

ulupo replied:
Done as you suggested

gtauzin commented on gtda/curves/features.py:
```diff
@@ -0,0 +1,217 @@
+                self.effective_function_params_ = self.function_params.copy()
+
+        else:
+            if isinstance(self.function, FunctionType):
```
Wouldn't that be a `elif` at the same level of indentation than the previous `else`?

ulupo replied:
I think you're right!

ulupo replied:
Changed.

gtauzin commented on gtda/curves/_functions.py:
```diff
@@ -0,0 +1,52 @@
+
+def _parallel_featurization(Xt, function, function_params, n_jobs):
+    if callable(function):
+        return function(Xt, axis=-1, **function_params)
```
I am not a big fan of having an edge case for the use of parallelism. Even if it is documented, it's likely that I would never read the `n_jobs` part of the doc and I doubt our users will. 

That being said, I think it's not such a big deal. xD

ulupo replied:
For what it's worth, I've added an explicit note in n_jobs saying that it is ignored when one of the strings is passed.

gtauzin replied:
I will be compromising here, although parallel on sliced arrays could also have been used so that `n_jobs` is always consistently used xD

ulupo replied:
Haha, ok.  Just academically though, I had a different impression on this issue.  If the user writes down an arbitrary function which works on 1d arrays but will break on 3d or 2d arrays, then slicing isn't an option right?

ulupo replied:
So basically my point is that the slicing logic is compatible with the predefined functions because they are numpy array functions, but not really compatible with arbitrary user-defined functions which satisfy the documentation requirements of mapping 1d -> 1d or scalar.

gtauzin commented on gtda/curves/_functions.py:
```diff
@@ -0,0 +1,52 @@
+    else:  # Assume function is a list or tuple of functions or None
+        channel_idx = [i for i, f in enumerate(function) if f is not None]
+        n_samples = len(Xt)
+        index_tups = product(*map(range, Xt.shape[:-2]), channel_idx)
```
I miss the reason for the complexity here with the map and the tuple formulations. Would that not work to just have lists and do:

```
        Xt = Parallel(n_jobs=n_jobs)(
            delayed(function[c])(Xt[s, c], **function_params[c])
            for c in Xt.shape[1] for s in Xt.shape[0]
            )
```
and then some concatenate like we do usually?

ulupo replied:
Ah yes you're right to note that. I wrote a version that worked on nd arrays just in case the discussion on PersistenceLandscape was still on the table. I agree though that if we settle on this not taking 4D arrays or higher, we should simplify the logic.

gtauzin replied:
I am not sure I follow. Even if the curve as 4D, all that should be done is in the beginning of `transform` to have:

```
if Xt.ndim == 4:
        Xt = Xt.reshape((Xt.shape[0], -1, Xt.shape[-1]))
```

And then it would be treated as all multi channel curves, no?

ulupo replied:
I don't think the reshape you suggest is equivalent to what I had in mind.  For instance I don't think it would return what we would want in the case of `PersistenceLandscape`, because calculating the maxes of each layer and then flattening the result is not the same as concatenating the layers first and then computing the maxes for each thus-flattened homology dimension (which I believe would result from your suggestion).

When writing those lines I had in mind an even greater level of generality in which one passes an n-d array and gets out an (n-1)-d array or a 2D array depending on whether a `flatten` parameter is passed as `True` or not. But we are converging on it being out of scope for this transformer, so I am not seriously defending any of this!

gtauzin replied:
`curves.StandardFeatures` has to take in curves and return features. If there is a concrete interest in having a functional transformer that `sklearn.preprocessing.FunctionTransformer` does not cover, then, I suggest we could have a `curves.StandardFunctionTransformer`.

But TBH, I really wonder what anyone would do with a 3D arrays of persistent landscape layers maxes, apart from actually reshaping it to use them as features :). On transfromers, we should give options to users that makes sense and for which we provide support in the form of a family of next transformers that will handle meaningfully the data structure outputted.

ulupo replied:
I agree with you that the extra generality is neither necessary nor appropriate here.  I just hadn't got round to committing the changes to avoid the `combinations` gymnastics.  I'm doing it now :)

gtauzin replied:
Ouffff :) Thanks! xD

ulupo replied:
The logic is now much simpler and tied to the input being always 3d.

gtauzin reviewed (commented):
Hi @ulupo. Thank you for your patience :)

A few comments here and there. It seems to me that the code would be simpler if `_function` and `_function_params`. Please let me know if I missed something.

ulupo committed:
Wrap up documentation

ulupo committed:
Improve indentation following @gtauzin's comment

ulupo committed:
Simplify iterations in Parallel call following @gtauzin's review

ulupo committed:
Simplify definition of effective_function_params_ following @gtauzin's review

ulupo committed:
Use more explicit shapes

ulupo committed:
Merge branch 'master' into curves

ulupo committed:
Add effective_function_ attribute

ulupo committed:
Raise deadlines for mapper tests

ulupo committed:
Merge branch 'master' into curves

ulupo committed:
Remove notebook tests even in macOS jobs

ulupo committed:
Add tests for attributes of fitted StandardFeatures

ulupo committed:
Linting

ulupo committed:
Simplify transform logic in PersistenceLandscape

ulupo committed:
Revert "Simplify transform logic in PersistenceLandscape"

This reverts commit 66f0d44e91b2c8b9c9411579bddcf9a394c5f6ee.

ulupo committed:
Linting

ulupo committed:
Make reshape more explicit

ulupo committed:
Make PersistenceLandscape.transform docstring more direct
